src/graph/graphcontext.c

```diff
@@ -0,0 +1,196 @@
+}
+
+RedisModuleKey* GraphContext_Key(RedisModuleCtx *ctx, const char *graph_name) {
+  int keylen = strlen(graph_name) + 4;
+  char strKey[keylen + 1];
+  sprintf(strKey, "%s_ctx", graph_name);
+
+  RedisModuleString *rm_graphctx = RedisModule_CreateString(ctx, strKey, keylen);
+  RedisModuleKey *key = RedisModule_OpenKey(ctx, rm_graphctx, REDISMODULE_WRITE);
+  RedisModule_FreeString(ctx, rm_graphctx);
+
+  return key;
+}
+
+GraphContext* GraphContext_New(RedisModuleCtx *ctx, RedisModuleString *rs_name) {
+  // Create key for GraphContext
+  const char *graph_name = RedisModule_StringPtrLen(rs_name, NULL);
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  if (RedisModule_ModuleTypeGetType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+    // Return NULL if key already exists
+    RedisModule_CloseKey(key);
+    return NULL;
+  }
+
+  GraphContext *gc = rm_malloc(sizeof(GraphContext));
+  gc->relation_cap = GRAPH_DEFAULT_RELATION_CAP;
+  gc->relation_count = 0;
+  gc->label_cap = GRAPH_DEFAULT_LABEL_CAP;
+  gc->label_count = 0;
+  gc->index_cap = 4;
+  gc->index_count = 0;
+
+  gc->graph_name = rm_strdup(graph_name);
+  gc->node_stores = rm_malloc(gc->label_cap * sizeof(LabelStore*));
+  gc->relation_stores = rm_malloc(gc->relation_cap * sizeof(LabelStore*));
+  gc->indices = rm_malloc(gc->index_cap * sizeof(Index*));
+
+  gc->node_allstore = LabelStore_New("ALL", GRAPH_NO_LABEL);
+  gc->relation_allstore = LabelStore_New("ALL", GRAPH_NO_RELATION);
+
+  // Set and close GraphContext in keyspace
+  RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+  RedisModule_CloseKey(key);
+
+  gc->g = Graph_New(GRAPH_DEFAULT_NODE_CAP);
+  // Add new Graph to keyspace
+  // TODO this separation feels rather clunky
+  key = RedisModule_OpenKey(ctx, rs_name, REDISMODULE_WRITE);
+  assert(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY);
+  RedisModule_ModuleTypeSetValue(key, GraphRedisModuleType, gc->g);
+  RedisModule_CloseKey(key);
+
+  return gc;
+}
+
+GraphContext* GraphContext_Get(RedisModuleCtx *ctx, RedisModuleString *rs_name) {
+  const char *graph_name = RedisModule_StringPtrLen(rs_name, NULL);
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  if (RedisModule_ModuleTypeGetType(key) != GraphContextRedisModuleType) return NULL;
+
+  GraphContext *gc = RedisModule_ModuleTypeGetValue(key);
+  RedisModule_CloseKey(key);
+
+  gc->g = Graph_Get(ctx, rs_name);
+
+  return gc;
+}
+
+// TODO next two functions are nearly identical; maybe consolidate
+LabelStore* GraphContext_AddLabel(GraphContext *gc, const char *label) {
+  if(gc->label_count == gc->label_cap) {
+    gc->label_cap += 4;
+    // Add space for additional label stores.
+    gc->node_stores = rm_realloc(gc->node_stores, gc->label_cap * sizeof(LabelStore*));
+    // Add space for additional label matrices.
+    // TODO is this too much of an abstraction breakage?
+    gc->g->_labels = rm_realloc(gc->g->_labels, gc->label_cap * sizeof(GrB_Matrix));
```

Comment:
If you think this approach makes sense, I'll just remove the `_` from `_labels`. Otherwise, we can pass a bool into `Graph_AddLabel` indicating that the array should be extended, or not have two separate functions...lots of options.